File: day2/cubeconundrum1.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <regex>

using namespace std;

int CheckGame(string& l, int& gamenum) {
    /* The code snippet is using regular expressions to search for matches in a string `l`. */
    smatch m;
    regex e ("\\d\\d\\s\\w+");
    int n, count;
    string str, color;



    while(regex_search(l, m, e)) {
        str = m[0].str();
        n = str.size() - 3;

        count = stoi(str.substr(0,2));
        color = str.substr(3,n);

        if (count > 14 || (color == "red" && count > 12) || (color == "green" && count > 13) || (color == "blue" && count > 14)) {
            return 0;
        }
            

        l = m.suffix().str();
    }

    return gamenum;
}


int main() {
    fstream f("input.txt");
    string line;
    int sum = 0, gamenum = 0;

    while(f.is_open() &&  getline(f, line)) {
        cout << line << endl;
        sum += CheckGame(line, ++gamenum); 
    }

    cout << "Sum of plausible game #s: " << sum << endl;

    f.close();

    return 0;
}

File: day2/rust/src/cubeconundrum1.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use regex::Regex;
use substring::Substring;

fn main() {
    let re = Regex::new(r"\d\d\s\w+").unwrap();
    let mut sum = 0;
    let mut i = 0;
    let mut gamenum = 0;

    if let Ok(lines) = read_lines("../input.txt") {
        for line in lines {
            if let Ok(l) = line {
                i += 1;
                gamenum = i;
                for cap in re.captures_iter(l.as_str()) {
                    let s = cap.get(0).unwrap().as_str();
                    let n = s.len();

                    let count = s.substring(0,2).parse::<usize>().unwrap();
                    let color = s.substring(3, n);

                    if count > 14 || (count > 12 && color == "red") || (count > 13 && color == "green") || (count > 14 && color == "blue") {
                        gamenum = 0;
                        break;
                    } 
                        
                }
                sum += gamenum;
            }
        }

        println!("Total game sum: {}", sum);
    }

}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}